3_visualize/src/plot_nhdv2_attr.R

```diff
@@ -0,0 +1,134 @@
+#' @param sites_epsg integer indicating the epsg code in the sites table 
+#' (i.e., 4269 for NAD83).
+#'
+#' @returns 
+#' Saves a png file containing a violin plot showing the distribution of each 
+#' NHDv2 attribute variable and returns the file path of the saved file. 
+#' 
+plot_nhdv2_attr <- function(attr_data,
+                            network_geometry,
+                            save_dir,
+                            plot_sites = FALSE, 
+                            sites = NULL, 
+                            sites_epsg = NULL){
+
+  if(plot_sites){
+    if (is.null(sites)){
+      stop('sites must be specified when plot_sites = TRUE')
+    }
+    if (is.null(sites_epsg)){
+      stop('sites CRS (as epsg code) must be specified when plot_sites = TRUE')
+    }
+    # Create spatial dataframe
+    sites_sf <- sf::st_as_sf(sites, coords = c('lon', 'lat'), crs = sites_epsg) 
+    
+    # add indicator to attr_data for reaches that have sites
+    attr_data_ind <- attr_data %>%
+      mutate(site_reaches = case_when(COMID %in% sites_sf$COMID ~ 1,
+                                      TRUE ~ 0))
+  }
+  
+  message("Plotting individual NHDv2 attribute variables...")
+  
+  plot_names <- vector('character', length = 0L)
+  
+  # For each column/attribute variable, plot the distribution of the data 
+  # across all NHDPlusv2 segments
+  attr_names <- names(attr_data)[names(attr_data) != "COMID"]
+  
+  for(i in seq_along(attr_names)){
+    col_name <- attr_names[i]
+    subset_cols <- c("COMID", col_name, "site_reaches")
+    
+    if(plot_sites){
+      dat_subset <- attr_data_ind %>%
+        select(any_of(subset_cols))
+
+      # plot the distribution of attr values on a linear scale
+      attr_plot <- dat_subset %>%
+        ggplot(aes(x = "", y = .data[[col_name]])) + 
+        geom_violin(draw_quantiles = c(0.5)) +
+        geom_jitter(data = dat_subset[dat_subset$site_reaches == 0,],
+                    height = 0, 
+                    size = 0.5,
+                    color = "steelblue",
+                    alpha = 0.1, width = 0.2) +
+        geom_jitter(data = dat_subset[dat_subset$site_reaches == 1,],
+                    height = 0, 
+                    color = "red",
+                    alpha = 0.4, width = 0.2) +
+        labs(x="") + 
+        theme_bw() + 
+        theme(plot.margin = unit(c(0,0,0,0), "cm"))
+      
+      # plot the spatial variation
+      attr_plot_spatial <- dat_subset %>% 
+        mutate(COMID = as.integer(COMID)) %>%
+        left_join(.,network_geometry[,c("COMID","geometry")],by=c("COMID"="COMID")) %>%
+        sf::st_as_sf() %>%
+        ggplot() + 
+        geom_sf(aes(color=.data[[col_name]]), size = 0.3) + 
+        scale_color_viridis_c(option="plasma") + 
+        theme_bw() + 
+        theme(plot.margin = unit(c(0,0,0,2), "cm"),
+              axis.text.x = element_text(size = 6),
+              legend.title = element_text(size = 10)) +
+        geom_sf(data = sites_sf, size = 0.3)
```

Comment:
The black dots for the site locations on the map are kind of hard to see, what about a white dot with a black outline? I don't know, pretty minor though

Reply:
Yeah, I know what you mean! As a first cut, I tried to make these points a little bit bigger and also added a bit of transparency to the river flowlines. I think this makes the site locations more visible, let me know what you think about that.

Reply:
> > One thing that might be useful, either as a separate plot or combined with these is a figure that labels the sites, either on a bar chart or on a map.
> 
> Good point! I opted not to include site labels on the map because the map was really cluttered since the sites are concentrated in a relatively small area. Instead, I added an optional argument, `label_sites`, to the `plot_nhdv2_attr()` function that adds site labels to the red points on the violin plot. Here's the same %clay example (the user can also adjust the size of the labels):
> 
> ![CAT_CLAYAVE](https://user-images.githubusercontent.com/8785034/196983237-072b1a18-c922-481b-8cf2-fb05621a87e1.png)
> 
> So it adds some clutter but as a relatively quick cut, might help quickly ID which sites have relatively higher or lower values of a given attribute. Do you think this helps?

Yeah I see what you mean, I like the violin plot point labeling option. Even if it does add some clutter, the default could be no labels and we could just flip the switch to check and compare quickly